File: Julia/example2_rockingblock/dynamics/res_fun.jl
using RockingBlock
using OrdinaryDiffEq
using LinearAlgebra
using ForwardDiff

function res_fun(u, params, ODESettings)
    n = 2

    # Extract variables from u
    t2, x2_bar, t1, x1_bar, xi, H_bar = u[1], u[2:3], u[4], u[5:6], u[7], u[8]

    # Simulate phase flows
    f1(dx, x, xi, t) = f(dx, x, xi, t, params) # (phase 1)
    x0 = x1_bar
    if t1 == 0
        x1   = x0
        Phi1 = I(n)
        psi1 = zeros(n,1)
    else
        prob_1 = ODEProblem(f1, x0, (0.0, t1), xi)
        sol_1  = solve(prob_1, ODESettings["solver"], reltol=ODESettings["reltol"], abstol=ODESettings["abstol"], save_everystep=false)
        function f1_fd(z)
            _prob_1 = remake(prob_1, u0 = z[1:2], p = z[3])
            solve(_prob_1, ODESettings["solver"], reltol=ODESettings["reltol"], abstol=ODESettings["abstol"], save_everystep=false)[1:n, 2]
        end
        dz = ForwardDiff.jacobian(f1_fd, [x0; xi])
        x1   = sol_1(t1)
        Phi1 = dz[:,1:n]
        psi1 = dz[:,n+1]
    end
    f1_val = zeros(n,1)
    f1(f1_val, x1, xi, [])
 

    f2(dx, x, xi, t) = f(dx, x, xi, t, params) # (phase 2)
    x0 = x2_bar
    if t2 == 0
        x2   = x0
        Phi2 = I(n)
        psi2 = zeros(n,1)
    else
        prob_2 = ODEProblem(f2, x0, (0.0, t2), xi)
        sol_2 = solve(prob_2, ODESettings["solver"], reltol=ODESettings["reltol"], abstol=ODESettings["abstol"], save_everystep=false)
        function f2_fd(z)
            _prob_2 = remake(prob_2, u0 = z[1:2], p = z[3])
            solve(_prob_2, ODESettings["solver"], reltol=ODESettings["reltol"], abstol=ODESettings["abstol"], save_everystep=false)[1:n, 2]
        end
        dz = ForwardDiff.jacobian(f2_fd, [x0; xi])
        x2   = sol_2(t2)
        Phi2 = dz[:,1:n]
        psi2 = dz[:,n+1]
    end
    f2_val = zeros(n,1)
    f2(f2_val, x2, xi, [])

    # Compute reset maps
    delta_12_(x) = delta_12(x, params)
    x1_P = delta_12_(x1)
    D12  = ForwardDiff.jacobian(delta_12_, x1)

    # Compute events and anchor
    e_12_(x) = e_12(x, [], [], params)
    e12_val  = e_12_(x1)
    de12 = ForwardDiff.gradient(e_12_, x1)'
    anchor_(x) = anchor(x, [], [], params)
    a  = anchor_(x2)
    da = ForwardDiff.gradient(anchor_, x2)'

    # Compute first integral
    H1, gradH1 = H(x1_bar, params)

    r = [x2 - x1_bar;
         a;
         x1_P - x2_bar;
         e12_val;
         H1 - H_bar]

    # Compute Jacobian
    R2 = [hcat(f2_val, Phi2) ; hcat(da*f2_val, da*Phi2)]
    R1 = [hcat(D12*f1_val, D12*Phi1); hcat(de12*f1_val, de12*Phi1)]
    Xi2 = vcat(psi2, da*psi2)
    Xi1 = vcat(D12*psi1, de12*psi1)
    C = [hcat(zeros(n, 1), -I(n)); zeros(1, n+1)]
    #Om = zeros(n+1, n+1)
    Or = zeros(1, n+1)
    Oc = zeros(n+1, 1)
    h1 = hcat(0, gradH1')

    dr = [hcat(R2, C, Xi2, Oc); 
          hcat(C, R1, Xi1, Oc);
          hcat(Or, h1, 0, -1)]

    return r, dr
end

File: Matlab_Octave/example4_slip/dynamics/resFun.m
function [r,dr,Phi_T] = resFun(u,params,ode,varargin)

if nargin < 4
    computePhi_T = false;
else
    computePhi_T = varargin{1};
end

nS = 4;
nF = 5;

% u = [t3; x3; t2; x2; t1; x1; xi; H_bar];
idxOS  = 0; % index offset
t3     = u(idxOS + 1);
idxOS  = idxOS + 1;
x3_bar = u(idxOS + (1:nS));
idxOS  = idxOS + nS;
t2     = u(idxOS + 1);
idxOS  = idxOS + 1;
x2_bar = u(idxOS + (1:nF));
idxOS  = idxOS + nF;
t1     = u(idxOS + 1);
idxOS  = idxOS + 1;
x1_bar = u(idxOS + (1:nS));
idxOS  = idxOS + nS;
xi     = u(idxOS + 1);
idxOS  = idxOS + 1;
H_bar  = u(idxOS + 1);

% simulate phase flows
n = nS;
f1_ode = @(t,x) f_S(x,xi,params);
X0 = [x1_bar;reshape(eye(n),[n^2,1]);zeros(n,1)]; 
if t1==0
    Xt1=X0;
else
    [~, Xt] = ode.solver(f1_ode, [0, t1], X0, ode.settings);
    Xt1 = Xt(end,:)';
end
x1   = Xt1(1:n);
Phi1 = reshape(Xt1(n+(1:n^2)),[n,n]);
psi1 = Xt1(n+n^2+(1:n));
f1   = f1_ode([],x1);

n = nF;
f2_ode = @(t,x) f_F(x,xi,params);
X0 = [x2_bar;reshape(eye(n),[n^2,1]);zeros(n,1)];
if t2==0
    Xt2=X0;
else
    [~, Xt] = ode.solver(f2_ode, [0, t2], X0, ode.settings);
    Xt2 = Xt(end,:)';
end
x2   = Xt2(1:n);
Phi2 = reshape(Xt2(n+(1:n^2)),[n,n]);
psi2 = Xt2(n+n^2+(1:n));
f2   = f2_ode([],x2);

n = nS;
f3_ode = @(t,x) f_S(x,xi,params);
X0 = [x3_bar;reshape(eye(n),[n^2,1]);zeros(n,1)]; 
if t3==0
    Xt3=X0;
else
    [~, Xt] = ode.solver(f3_ode, [0, t3], X0, ode.settings);
    Xt3 = Xt(end,:)';
end
x3   = Xt3(1:n);
Phi3 = reshape(Xt3(n+(1:n^2)),[n,n]);
psi3 = Xt3(n+n^2+(1:n));
f3   = f3_ode([],x3);

% compute reset maps
[x1_P,D12] = delta_lo(x1, params);
[x2_P,D23] = delta_td(x2, params);

% compute events and anchor
[e12, ~, ~, de12] = e_lo(x1, params);
[e23, ~, ~, de23] = e_td(x2, params);
[a, ~, ~, da]     = anchor(x3, params);

% compute first integral
[H1,dH1,~] = H_S(x1_bar,params);

r  = [x3-x1_bar;
      a;
      x2_P-x3_bar;
      e23;
      x1_P-x2_bar;
      e12;
      H1-H_bar];

if computePhi_T
    % compute saltation matrices for monodromy matrix
    f2P = f2_ode([],x2_bar);
    S12 = D12 + (f2P-D12*f1)/(de12*f1);
    
    f3P = f3_ode([],x3_bar);
    S23 = D23 + (f3P-D23*f2)/(de23*f2);
    
    Phi_T = Phi3*S23*Phi2*S12*Phi1;
else
    Phi_T = [];
end

% compute Jacobian
R3  = [f3,Phi3;da*f3,da*Phi3];
R2  = [D23*f2,D23*Phi2;de23*f2,de23*Phi2];
R1  = [D12*f1,D12*Phi1;de12*f1,de12*Phi1];
Xi3 = [psi3;da*psi3];
Xi2 = [D23*psi2;de23*psi2];
Xi1 = [D12*psi1;de12*psi1];

CF  = [zeros(nF,1),-eye(nF);0,zeros(1,nF)];
CS  = [zeros(nS,1),-eye(nS);0,zeros(1,nS)];
OSF = zeros(nS+1,nF+1);   % zero matrix
OFS = zeros(nF+1,nS+1);   % zero matrix
OSS = zeros(nS+1,nS+1);   % zero matrix
%OFF = zeros(nF+1,nF+1);   % zero matrix
OcF  = zeros(nF+1,1); % zero column
OcS  = zeros(nS+1,1); % zero column
OrF  = zeros(1,nF+1); % zero row
OrS  = zeros(1,nS+1); % zero row
h1  = [0,dH1];

dr = [R3  OSF CS  Xi3 OcS;
      CS  R2  OSS Xi2 OcS;
      OFS CF  R1  Xi1 OcF;
      OrS OrF h1  0   -1];
end

